src/drivers/amd/common/acp_dma.c

```diff
@@ -0,0 +1,396 @@
+						 unsigned int req_chan)
+{
+	k_spinlock_key_t key;
+	struct dma_chan_data *channel;
+
+	key = k_spin_lock(&dma->lock);
+	if (req_chan >= dma->plat_data.channels) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acpdma_tr, "DMA: Channel %d not in range", req_chan);
+		return NULL;
+	}
+	channel = &dma->chan[req_chan];
+	if (channel->status != COMP_STATE_INIT) {
+		k_spin_unlock(&dma->lock, key);
+		tr_err(&acpdma_tr, "DMA: channel already in use %d", req_chan);
+		return NULL;
+	}
+	atomic_add(&dma->num_channels_busy, 1);
+	channel->status = COMP_STATE_READY;
+	k_spin_unlock(&dma->lock, key);
+	/* reset read and write pointers */
+	struct acp_dma_chan_data *acp_dma_chan = dma_chan_get_data(channel);
+
+	acp_dma_chan->config[req_chan].rd_size = 0;
+	acp_dma_chan->config[req_chan].wr_size = 0;
+	return channel;
+}
+
+static void acp_dma_channel_put(struct dma_chan_data *channel)
+{
+	k_spinlock_key_t key;
+	struct acp_dma_chan_data *acp_dma_chan = dma_chan_get_data(channel);
+
+	key = k_spin_lock(&channel->dma->lock);
+	channel->status = COMP_STATE_INIT;
+	atomic_sub(&channel->dma->num_channels_busy, 1);
+	k_spin_unlock(&channel->dma->lock, key);
+	/* reset read and write pointer */
+	acp_dma_chan->config[channel->index].rd_size = 0;
+	acp_dma_chan->config[channel->index].wr_size = 0;
+}
+
+/* Stop the requested channel */
+static int acp_dma_stop(struct dma_chan_data *channel)
+{
+	acp_dma_cntl_0_t	 dma_cntl;
+	acp_dma_ch_sts_t	 ch_sts;
+	uint32_t dmach_mask;
+	uint32_t delay_cnt = 10000;
+
+	switch (channel->status) {
+	case COMP_STATE_READY:
+	case COMP_STATE_PREPARE:
+		return 0; /* do not try to stop multiple times */
+	case COMP_STATE_PAUSED:
+	case COMP_STATE_ACTIVE:
+		break;
+	default:
+		return -EINVAL;
+	}
+	channel->status = COMP_STATE_READY;
+	dmach_mask = (1 << channel->index);
+	dma_cntl = (acp_dma_cntl_0_t)dma_chan_reg_read(channel, ACP_DMA_CNTL_0);
+	/* Do the HW stop of the DMA */
+	/* set DMAChRst bit to stop the transfer */
+	dma_cntl.bits.dmachrun      = 0;
+	dma_cntl.bits.dmachiocen    = 0;
+	dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+	ch_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+	if (ch_sts.bits.dmachrunsts & dmach_mask) {
+		/* set the reset bit for this channel to stop the dma transfer */
+		dma_cntl.bits.dmachrst	 = 1;
+		dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+	}
+
+	while (delay_cnt > 0) {
+		ch_sts = (acp_dma_ch_sts_t)dma_reg_read(channel->dma, ACP_DMA_CH_STS);
+		if (!(ch_sts.bits.dmachrunsts & dmach_mask)) {
+			/* clear the reset flag after successfully stopping the dma transfer
+			 * and break from the loop
+			 */
+			dma_cntl.bits.dmachrst = 0;
+			dma_chan_reg_write(channel, ACP_DMA_CNTL_0, dma_cntl.u32all);
+			break;
+		}
+		delay_cnt--;
+	}
+	return 0;
+}
+
+static int acp_dma_start(struct dma_chan_data *channel)
+{
+	acp_dma_cntl_0_t dma_cntl;
+	acp_dma_ch_sts_t dma_sts;
+	uint32_t chan_sts;
+	int ret = 0;
```

Comment:
not needed, just do `return 0` in the end of the function

Reply:
Done